### ISSUE TITLE
Make private inputs explicit

### ISSUE THREAD

**Issue by Schaeff** [JacobEberhardt/ZoKrates#13]
# Problem
The way private inputs are treated is different from public ones:
- Private inputs are implicit, public inputs are explicit:
```
def main(c):
  c == a * b
  return 1
```
- Public inputs are passed as arguments in `compute-witness`, while private inputs are passed interactively
- It makes it hard to use `compute-witness` programatically
- It gives no way to list private inputs without solving the R1CS

# Solution
We think they should be explicit:
```
def main(c | a,b):
  c == a * b
  return 1
```

# Roadmap

1. Make it possible to pass private inputs as arguments in `compute-witness` rather than interactively (the interactive mode should be activated by a flag)
2. Make it possible to specify explicit private inputs in `.code` files
3. Ban implicit private inputs from `.code` files

Thoughts welcome :)

**Comment by JacobEberhardt** [comment]
We'll just leave this here for a bit to collect input!

**Comment by nicola** [comment]
I like the idea, it's just that it looks kind of odd syntax.

What if you require main to have two only: `(x, w)` which are vectors of inputs? (I know this complicates syntax).

**Comment by Schaeff** [comment]
Yes I guess the syntax is up for debate. Vectors are not supported elsewhere so that could be confusing. Other ideas:

- `(a,b)(c,d)` unless higher order functions are on the roadmap
- `(private a, public b, private c)`
- other?

Also, I wonder if that should be defined only for `main`, or all functions. It doesn't sound necessary for other functions, however it could introduce a consistent notion of scope.

**Comment by nicola** [comment]
I like `(private a, public b, private c)`, but I also like the experimental syntax

**Comment by PeterBorah** [comment]
+1 to some sort of explicit syntax. `(a, b, private c, private d)` is probably the most obvious/conservative choice, though I am ok with the `|` notation and the `()()` notation as well.

**Comment by Schaeff** [comment]
`(a, b, private c, private d)` is being implemented in #15

**Comment by Schaeff** [comment]
Merged into develop
